torch/distributed/device_mesh.py

```diff
@@ -197,9 +197,12 @@ def __init__(
             *,
             mesh_dim_names: Optional[Tuple[str, ...]] = None,
         ) -> None:
+            if isinstance(mesh, torch.Tensor) and mesh.get_device() != -1:
```

Comment:
nit: I think it should be `mesh.device.type != "cpu"`

Reply:
maybe just assert with assertion message should be enough ?

Reply:
IMHO, I think `ValueError` is good.

One idea: We can prefer to raise real errors (i.e. not `assert ...`) if it is user facing. Practically for PyTorch, `assert cond, msg` and `if not cond: raise AssertionError(msg)` are the same, but distinguishing them can be one way to indicate what is an internal assert vs. intended to be user-facing.

Reply:
@wanchaol I think Andrew raised a good point here. Raising a more explicit error does give users a little bit more information about what they did wrong. IMO, I prefer raising the ValueError to let users know they have entered invalid arg for DeviceMesh. Wdyt?

Reply:
Yeah sounds legit to me!